### PR TITLE
tier-2_rados_basic_regression: Verification of Fragmentation scores

### PR DESCRIPTION
**Pull request by pdhiran**
Verification of Fragmentation Scores & Allocation size.

Steps:
1. Create a test pool, read/write/delete data, to create fragmentation on OSDs.
2. Collect Sample set of OSDs to test the allocation size and fragmentation scores.
3. Checking the config param for allocation size for ssd & hdd.
4. Checking the allocation size from the bluestore dump.
5. Checking the fragmentation scores.
6. Do this on all sample OSDs collected.

Signed-off-by: Pawan Dhiran <pdhiran@redhat.com>

### CODE REVIEW

**Comment by pdhiran** [comment]
Pass logs : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-X8TSPA

File: suites/pacific/rados/tier-2_rados_test-pool-functionalities.yaml [hunk 1]
```diff
@@ -291,3 +291,10 @@ tests:
           pool_name: pool_num_chk
           delete_pool: true
 
+  - test:
+      name: OSD min-alloc size & fragmentation checks
```
Comment (pdhiran) [inline suites/pacific/rados/tier-2_rados_test-pool-functionalities.yaml:295]: Remove &, which might be treated as wild char. use "and"

File: suites/quincy/rados/tier-2_rados_test-pool-functionalities.yaml [hunk 2]
```diff
@@ -275,3 +275,11 @@ tests:
         verify_pg_num_limit:
           pool_name: pool_num_chk
           delete_pool: true
+
+  - test:
+      name: OSD min-alloc size & fragmentation checks
```
Comment (pdhiran) [inline suites/quincy/rados/tier-2_rados_test-pool-functionalities.yaml:280]: Remove &, which might be treated as wild char. use "and"

File: tests/rados/rados_prep.py [hunk 3]
```diff
@@ -173,5 +177,78 @@ def run(ceph_cluster, **kw):
...
+        pool_id = pool_obj.get_pool_id(pool_name=pool_name)
+        pgid = f"{pool_id}.0"
+        pg_set = rados_obj.get_pg_acting_set(pg_num=pgid)
+
+        log.info(f"Checking the fragmentation & min-alloc scores on OSDs : {pg_set}")
+        for osd in pg_set:
+            host = rados_obj.fetch_host_node(daemon_type="osd", daemon_id=str(osd))
+            if not host:
+                log.error("failed to find host for the osd")
+                return 1
```
Comment (pdhiran) [inline tests/rados/rados_prep.py:198]: Raise exception.

File: tests/rados/rados_prep.py [hunk 4]
```diff
@@ -173,5 +177,78 @@ def run(ceph_cluster, **kw):
...
+            params = ["bluestore_min_alloc_size_hdd", "bluestore_min_alloc_size_ssd"]
+            for param in params:
+                alloc_cmd = (
+                    f"ceph daemon /var/run/ceph/ceph-osd.{osd}.asok config get {param}"
+                )
+                cmd = f"cephadm -v shell -- {alloc_cmd} -f json"
+                log.debug(f"Checking the {param} on {osd} on host {host.hostname}.")
+                out, err = host.exec_command(sudo=True, cmd=cmd)
+                details = json.loads(out)
+                if int(details[param]) != 4096:
```
Comment (pdhiran) [inline tests/rados/rados_prep.py:210]: Pass the allocation size from the conf file. Make it parametrized.

File: tests/rados/rados_prep.py [hunk 5]
```diff
@@ -173,5 +177,78 @@ def run(ceph_cluster, **kw):
...
+                    )
+                    return 1
+
+            # Checking min-alloc size for the OSD from the dump
+            alloc_cmd = f"ceph daemon /var/run/ceph/ceph-osd.{osd}.asok bluestore allocator dump block"
+            cmd = f"cephadm -v shell -- {alloc_cmd} -f json"
+            log.debug(f"Checking the allocation size on {osd} on host {host.hostname}.")
+            out, err = host.exec_command(sudo=True, cmd=cmd)
+            details = json.loads(out)
+            if int(details["alloc_unit"]) != 4096:
```
Comment (pdhiran) [inline tests/rados/rados_prep.py:222]: same as above.

File: tests/rados/rados_prep.py [hunk 6]
```diff
@@ -173,5 +177,78 @@ def run(ceph_cluster, **kw):
...
+                log.error(
+                    f"Allocation unit for osd {osd} is not 4k, it is {details['alloc_unit']}"
+                )
+                return 1
```
Comment (pdhiran) [inline tests/rados/rados_prep.py:226]: raise exception.

File: tests/rados/rados_prep.py [hunk 7]
```diff
@@ -173,5 +177,78 @@ def run(ceph_cluster, **kw):
...
+            log.info(
+                f"Fragmentation score for the OSD : {details['fragmentation_rating']}"
+            )
+
+            if 0.9 < float(details["fragmentation_rating"]) < 1.0:
+                log.error(
+                    f"Fragmentation on osd {osd} is dangerously high."
+                    f"Ideal range 0.0 to 0.7. Actual fragmentation on OSD: {details['alloc_unit']}"
+                )
+                return 1
```
Comment (pdhiran) [inline tests/rados/rados_prep.py:245]: Raise exception.

File: suites/pacific/rados/tier-2_rados_test-pool-functionalities.yaml (hunk 1 above)
Comment (pdhiran) [reply suites/pacific/rados/tier-2_rados_test-pool-functionalities.yaml:295]: Done

File: suites/quincy/rados/tier-2_rados_test-pool-functionalities.yaml (hunk 2 above)
Comment (pdhiran) [reply suites/quincy/rados/tier-2_rados_test-pool-functionalities.yaml:280]: Done

File: tests/rados/rados_prep.py (hunk 3 above)
Comment (pdhiran) [reply tests/rados/rados_prep.py:198]: Done

File: tests/rados/rados_prep.py (hunk 4 above)
Comment (pdhiran) [reply tests/rados/rados_prep.py:210]: Done

File: tests/rados/rados_prep.py (hunk 5 above)
Comment (pdhiran) [reply tests/rados/rados_prep.py:222]: Done

File: tests/rados/rados_prep.py (hunk 6 above)
Comment (pdhiran) [reply tests/rados/rados_prep.py:226]: Done

File: tests/rados/rados_prep.py (hunk 7 above)
Comment (pdhiran) [reply tests/rados/rados_prep.py:245]: Done.

**Comment by pdhiran** [comment]
@sunilkumarn417 , Addressed Changes. ready for review.

**Review by sunilkumarn417 — APPROVED** [review]

**Pull request merged**